orttraining/orttraining/training_ops/cpu/communication/recv.cc

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/training_ops/cpu/communication/recv.h"
+
+#include <mpi.h>
+
+#include "orttraining/training_ops/communication_common.h"
+#include "orttraining/core/framework/mpi_context.h"
+
+namespace onnxruntime {
+namespace contrib {
+
+void Recv::ReceiveData(
+    const int num_tensors,
+    std::vector<Tensor*> received_tensors,
+    const int src,
+    const size_t aggregated_aligned_tensor_bytes,
+    std::vector<char>& buffer) const {
+  buffer.reserve(aggregated_aligned_tensor_bytes);
+  CommInfo_t info_data{buffer.data(),
+                       static_cast<int>(aggregated_aligned_tensor_bytes),
+                       src,
+                       static_cast<int>(tag_)};
+
+  MPI_CHECK(MPI_Recv(
+      info_data.buffer, info_data.size, MPI_CHAR,
+      info_data.rank, info_data.tag, MPI_COMM_WORLD, MPI_STATUS_IGNORE));
+
+  // Copy tensors from buffer to outputs.
+  size_t tensor_offset_in_bytes = 0;
+  for (int i = 0; i < num_tensors; ++i) {
+    Tensor* tensor = received_tensors[i];
+
+    // Find the next aligned offset in the tensor buffer to meet alignment requirement
+    tensor_offset_in_bytes = GetAggregatedAlignedAddress(tensor_offset_in_bytes);
+
+    // Copy data out from buffer.
+    memcpy(tensor->MutableDataRaw(), buffer.data() + tensor_offset_in_bytes, tensor->SizeInBytes());
+
+    tensor_offset_in_bytes += tensor->SizeInBytes();
+  }
+}
```

Comment:
Is it correct to assert here that tensor_offset_in_bytes matches aggregated_aligned_tensor_bytes?  i.e., that after extracting the tensors, the position we reach in the block of data received matches up with the end of the block?